### PR TITLE
(SIMP-6740) Add tests for service toggles

### PR DESCRIPTION
**Pull request by trevor-vaughan**
Added tests to flip:
  * FIPS mode
  * SELinux Mode
  * Auditd

Wait for 10 minutes when bootstrapping the server by default in case
we're running on an overloaded system.

Requires the following to work:
  * simp/pupmod-simp-simplib#228
  * simp/pupmod-simp-pam#94

Note: Currently pointing at trevor-vaughan/beaker until the patch in
fix_reboot_check_fix can be merged and released. The acceptance tests
will probably fail without this patch.

SIMP-6740 #close

### CODE REVIEW

File: Puppetfile.pinned [hunk 1]
```diff
@@ -348,7 +348,7 @@ mod 'simp-simpkv',
 
 mod 'simp-simplib',
   :git => 'https://github.com/simp/pupmod-simp-simplib',
-  :tag => '4.2.0'
+  :branch => 'master'
```
Comment (op-ct) [inline Puppetfile.pinned:351]: Are we unpinning all these branches (instead of pinning them to new versions) because we still expect a lot of flux?

File: Gemfile [hunk 2]
```diff
@@ -1,37 +1,51 @@
...
+  ENV['EXTRA_GEMFILE'] || '',
+  "#{__FILE__}.project",
+  "#{__FILE__}.local",
+  File.join(Dir.home, '.gemfile'),
+]
+extra_gemfiles.each do |gemfile|
+  if File.file?(gemfile) && File.readable?(gemfile)
+    eval(File.read(gemfile), binding)
+  end
+end
```
Comment (op-ct) [inline Gemfile:51]: ✨ :+1: ✨

**Comment by op-ct** [comment]
From the failing test at https://travis-ci.com/github/simp/simp-core/jobs/377941564#L1273:

```
el6-build-server 13:25:00$ runuser build_user -l -c  'cd /home/travis/build/simp/simp-core; bundle exec rake deps:checkout'
  WARNING: cannot load such file -- simp/build/release_mapper
  rake aborted!
  Don't know how to build task 'deps:checkout' (See the list of available tasks with `rake --tasks`)
  Did you mean?  deps:clean
  /simp-core/vendor/ruby/2.4.0/gems/rake-12.3.3/exe/rake:27:in `<top (required)>'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/cli/exec.rb:63:in `load'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/cli/exec.rb:63:in `kernel_load'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/cli/exec.rb:28:in `run'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/cli.rb:476:in `exec'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/cli.rb:30:in `dispatch'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/cli.rb:24:in `start'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/exe/bundle:46:in `block in <top (required)>'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'
  /home/build_user/.rvm/gems/ruby-2.4.5/gems/bundler-2.1.1/exe/bundle:34:in `<top (required)>'
  /home/build_user/.rvm/gems/ruby-2.4.5/bin/bundle:23:in `load'
  /home/build_user/.rvm/gems/ruby-2.4.5/bin/bundle:23:in `<main>'
  /home/build_user/.rvm/gems/ruby-2.4.5/bin/ruby_executable_hooks:24:in `eval'
  /home/build_user/.rvm/gems/ruby-2.4.5/bin/ruby_executable_hooks:24:in `<main>'
  (See full trace by running task with --trace)
```

File: Puppetfile.pinned (hunk 1 above)
Comment (trevor-vaughan) [reply Puppetfile.pinned:351]: Yeah, this seems to be the focus of most changes right now.

File: spec/acceptance/suites/default/87_disable_fips_spec.rb [hunk 3]
```diff
@@ -40,7 +40,7 @@
         ensure_ssh_connection(agent)
 
         retry_on(agent, 'puppet agent -t',
-          :desired_exit_codes => [0],
+          :desired_exit_codes => [0,2],
```
Comment (lnemsick-simp) [inline spec/acceptance/suites/default/87_disable_fips_spec.rb:43]: Why was this changed to allow exit code 2 (changes detected)?

File: spec/acceptance/suites/default/87_disable_fips_spec.rb [hunk 4]
```diff
@@ -69,13 +69,12 @@
         retry_on(agent, 'uptime', :retry_interval => 15 )
 
         retry_on(agent, 'puppet agent -t',
-          :desired_exit_codes => [0],
+          :desired_exit_codes => [0,2],
```
Comment (lnemsick-simp) [inline spec/acceptance/suites/default/87_disable_fips_spec.rb:72]: Why was this changed to allow exit code 2 (changes detected)?  Shouldn't we be testing that it converges to a stable configuration?

File: spec/acceptance/suites/default/85_disable_selinux_spec.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    it 'should apply non-selinux-mode settings on agents' do
+      # Wait for the puppetserver to be ready to receive requests
+      retry_on(master, puppetserver_status_cmd, :retry_interval => 10)
+
+      block_on(agents, :run_in_parallel => false) do |agent|
+        # Wait for the machine to come back up
+        retry_on(agent, 'uptime', :retry_interval => 15 )
+
+        retry_on(agent, 'puppet agent -t',
+          :desired_exit_codes => [0,2],
```
Comment (lnemsick-simp) [inline spec/acceptance/suites/default/85_disable_selinux_spec.rb:71]: Why do we allow exit code 2 (changes detected)? Shouldn't we be iterating until no changes have been detected?

File: spec/acceptance/suites/default/89_disable_audit_spec.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    it 'should undo compliance-enforced settings on agents' do
+      block_on(agents, :run_in_parallel => false) do |agent|
+        # FIXME: By the time we get to the last node, the ssh connection may have been
+        # aggressively terminated by beaker for that node instead of being handled with
+        # reconnect-after-timeout logic.
+        ensure_ssh_connection(agent)
+
+        # We need to do this because the simplib__auditd fact will not be
+        # triggered until after the first puppet run which will cause PAM to
+        # deny access to the system due to the pam_tty_audit bug.
```
Comment (lnemsick-simp) [inline spec/acceptance/suites/default/89_disable_audit_spec.rb:45]: I don't understand the comment

File: spec/acceptance/suites/default/89_disable_audit_spec.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    it 'should apply non-auditd-mode settings on agents' do
+      # Wait for the puppetserver to be ready to receive requests
+      retry_on(master, puppetserver_status_cmd, :retry_interval => 10)
+
+      block_on(agents, :run_in_parallel => false) do |agent|
+        # Wait for the machine to come back up
+        retry_on(agent, 'uptime', :retry_interval => 15 )
+
+        retry_on(agent, 'puppet agent -t',
+          :desired_exit_codes => [0,2],
```
Comment (lnemsick-simp) [inline spec/acceptance/suites/default/89_disable_audit_spec.rb:77]: Why do we allow exit code 2 (changes detected)? Shouldn't we be iterating until no changes have been detected?

File: spec/acceptance/suites/default/85_disable_selinux_spec.rb [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  context 'set up hiera to disable selinux' do
+    let(:prod_env_dir) { '/etc/puppetlabs/code/environments/production' }
+    let(:prod_env_default_yaml) { File.join(prod_env_dir, 'data', 'default.yaml') }
+
+    it 'should disable compliance enforcement and disable FIPS' do
+      hiera = YAML.load(on(master, "cat #{prod_env_default_yaml}").stdout)
+
+      # Set profile list to empty string in the existing hiera, because
+      # when no profiles are specified, there is nothing to enforce!
+      hiera['compliance_markup::enforcement'] = []
```
[the remark this answers is not in the document]
Comment (trevor-vaughan) [reply spec/acceptance/suites/default/85_disable_selinux_spec.rb:20]: Because I didn't backtrace through the entire stack to see if it would mess things up or not.

File: spec/acceptance/suites/default/87_disable_fips_spec.rb (hunk 3 above)
Comment (trevor-vaughan) [reply spec/acceptance/suites/default/87_disable_fips_spec.rb:43]: In case auditd toggles (it does sometimes)

Comment (trevor-vaughan) [reply spec/acceptance/suites/default/87_disable_fips_spec.rb:43]: It's not an idempotency check so it doesn't matter.

File: spec/acceptance/suites/default/87_disable_fips_spec.rb (hunk 4 above)
Comment (trevor-vaughan) [reply spec/acceptance/suites/default/87_disable_fips_spec.rb:72]: It's not an idempotency check and the reboot notification may trigger an exit code of  `2`

File: spec/acceptance/suites/default/85_disable_selinux_spec.rb (hunk 5 above)
Comment (trevor-vaughan) [reply spec/acceptance/suites/default/85_disable_selinux_spec.rb:71]: ☝️

File: spec/acceptance/suites/default/89_disable_audit_spec.rb (hunk 6 above)
Comment (trevor-vaughan) [reply spec/acceptance/suites/default/89_disable_audit_spec.rb:45]: 1. `puppet agent -t` => kill auditd
2. You can't SSH in any more because `pam_tty_audit` will kill the session
3. Work around it by running puppet twice in the same run

File: spec/acceptance/suites/default/89_disable_audit_spec.rb (hunk 7 above)
Comment (trevor-vaughan) [reply spec/acceptance/suites/default/89_disable_audit_spec.rb:77]: Ditto from above

File: spec/acceptance/suites/default/87_disable_fips_spec.rb (hunk 3 above)
Comment (lnemsick-simp) [reply spec/acceptance/suites/default/87_disable_fips_spec.rb:43]: IMHO  it would be better to leave the system in a stable state for the next test.

File: Gemfile [hunk 9]
```diff
@@ -1,37 +1,52 @@
...
   gem 'pry-byebug'
   gem 'pry-doc'
 end
 
-#vim: set syntax=ruby:
+group :system_tests do
+  #gem 'beaker'
+  # This is needed until beaker can get it merged and released upstream,
+  # otherwise the acceptance tests will probably fail
+  gem 'beaker', :git => 'https://github.com/trevor-vaughan/beaker', :branch => 'fix_reboot_check_fix'
```
Comment (op-ct) [inline Gemfile:36]: As noted in https://github.com/voxpupuli/beaker/pull/1677#discussion_r478729042, the logic of this beaker fix is fundamentally broken (and guts some beaker exceptions in the process)―we shouldn't expect it to merge upstream without extensive revision.  The changes also can't help these acceptance tests pass, unless it's through some kind of oblique side effect (perhaps buying some time in an undiagnosed race condition?).

Comment (trevor-vaughan) [reply Gemfile:36]: @op-ct I've left comments over there. Presently, it works (regardless of how it works) and unfortunately, I have to push on with other things. I tried all sorts of `sleep` calls first just to see if it was something that was easy to work around, but it wasn't.

Due to time constraints, this is what it is unless someone else has time to make it better.

File: spec/acceptance/suites/default/87_disable_fips_spec.rb (hunk 3 above)
Comment (trevor-vaughan) [reply spec/acceptance/suites/default/87_disable_fips_spec.rb:43]: All of these tests were pushed to the very end of the stack because, by their nature, they may cause issues with system stability.

The final test that shoves everything back into a compliant mode was not modified and should achieve stability.
